File: DotNetOnion/Helpers/CommandsHelper.cs
using DotNetOnion.Cells;
using System;
using System.Collections.Generic;
using System.Text;

namespace DotNetOnion.Helpers
{
    public static class CommandsHelper
    {
        public static bool IsVariableLength(byte command) =>
            command == 7 || command >= 128;

        public static bool IsRelayCell(byte command) =>
            command == 9 || command == 3;

        //FIXME: Maybe reflection is cleaner?
        public static Cell GetCell(byte command)
        {
            return command switch
            {
                0 => new CellPadding(),
                3 => new CellRelayEncrypted(),
                4 => new CellDestroy(),
                5 => new CellCreateFast(),
                6 => new CellCreatedFast(),
                7 => new CellVersions(),
                8 => new CellNetInfo(),
                129 => new CellCerts(),
                130 => new CellAuthChallenge(),
                _ => throw new NotImplementedException(),
            };
        }
    }
}

File: NOnion/Helpers/CommandsHelper.fs
namespace NOnion.Helpers

open System.IO

open NOnion.Cells

module CommandsHelper =
    [<Literal>]
    let CellAuthChallengeCommand = 130uy
    [<Literal>]
    let CellCertsCommand = 129uy
    [<Literal>]
    let CellVersionCommand = 7uy
    [<Literal>]
    let CellNetInfoCommand = 8uy
    [<Literal>]
    let CellCreateFastCommand = 5uy
    [<Literal>]
    let CellCreatedFastCommand = 6uy

    let IsVariableLength (command: byte): bool =
        command = 7uy || command >= 128uy

    let GetCell (command: byte) (reader: BinaryReader): ICell =
        match command with 
        | CellAuthChallengeCommand ->
            CellAuthChallenge.Deserialize reader
        | CellCertsCommand ->
            CellCerts.Deserialize reader
        | CellVersionCommand ->
            CellVersions.Deserialize reader
        | CellNetInfoCommand ->
            CellNetInfo.Deserialize reader
        | CellCreateFastCommand ->
            CellCreateFast.Deserialize reader
        | CellCreatedFastCommand ->
            CellCreatedFast.Deserialize reader
        | _ -> failwith "not implemented!"
